feat: add parameter description (#171)
autowarefoundation/autoware-documentation: docs/contributing/coding-guidelines/ros-nodes/parameters.md

# Parameters

> Autoware has following parameters

Reviewer: Better?
```suggestion
Autoware has the following types of parameters:
```

Author: https://github.com/autowarefoundation/autoware-documentation/pull/171/commits/38fb95acd07707485b634c4a8b01eed53ef5f2e1
DONE

- `p1`: reference parameter for the node package
  - e.g., [the parameter for the `behavior_path_planner` package](https://github.com/autowarefoundation/autoware.universe/tree/main/planning/behavior_path_planner/config)

> - `p2`: reference parameter for the module launch

Author: I referred tier4_*_launch to the module launch. I'm wondering about a better name.

  - e.g., [the parameter for the `tier4_planning_launch` package](https://github.com/autowarefoundation/autoware.universe/tree/main/launch/tier4_planning_launch/config/)

> - `p3`: [reference parameter for the `autoware_launch` package](https://github.com/autowarefoundation/autoware_launch/tree/main/autoware_launch/config)

Reviewer: How about converting this list to a table?
Also, I feel it's better to use more descriptive names for `p1`, `p2`, and `p3`.
For example, `Node Parameter`, `Component Parameter`, and `Launch Parameter`? :thinking:

Reviewer: Also, is it possible to add a figure to visually explain these types of parameters?

Author: Thank you for your good idea. I've fixed what I agreed.
The explanation is not short words but sentences, so I feel it better to use itemization rather than the table.

</include>
```

### When launching autoware via the module launch

In [`planning.launch.xml`](https://github.com/autowarefoundation/autoware.universe/blob/main/launch/tier4_planning_launch/launch/planning.launch.xml) of the `tier4_planning_launch` package, the default variable for the `tier4_planning_launch_param_path` argument is `$(find-pkg-share tier4_planning_launch)/config`.
Therefore, parameters in the `tier4_planning_launch` package (`p2`) are used.

```xml
<arg name="tier4_planning_launch_param_path" default="$(find-pkg-share tier4_planning_launch)/config" description="tier4_planning_launch parameter path"/>
```

## Automatic Synchronization of parameters

There is a PR generated automatically to synchronize `p3` with `p2`.

## How to manage parameters

We explain how to manage parameters in some use cases.

### When creating a new package

- Locates the parameters of `p1` and `p2` for the new package in `autowarefoundation/autoware.universe`.
- Create a PR to `autowarefoundation/autoware.universe`.
- Update [a synchronization setting file](TODO).
- Sync PR to `autowarefoundation/autoware_launch` is created automatically.

>   - Or you can create this PR manually.

Author: I forgot the detail about the sync, but is it ok to modify autoware_launch/config manually?

# Parameters

Autoware has the following types of parameters:

- `Node Parameter`: the reference parameter for the node package
  - e.g., [the parameter for the `behavior_path_planner` package](https://github.com/autowarefoundation/autoware.universe/tree/main/planning/behavior_path_planner/config)
- `Component Parameter`: the reference parameter for the component launch
  - e.g., [the parameter for the `tier4_planning_launch` package](https://github.com/autowarefoundation/autoware.universe/tree/main/launch/tier4_planning_launch/config/)

> - `Launch Parameter`: [the reference parameter for the `autoware_launch` package](https://github.com/autowarefoundation/autoware_launch/tree/main/autoware_launch/config)

Author: We have a system component, so `System Parameter` is confusing.

`Autoware Parameter` may be better?
@kenji-miyake What do you think?

`Launch Parameter` contains the same parameters as `Component Parameter`.
`Component Parameter` contains all the parameters of `Launch Parameter`, and also parameters of multiple node instances from a single node package with different parameters.

![parameter-architecture](images/parameter-architecture.svg){: style="width:900px"}

## Parameters to be used

### When launching autoware via the `autoware_launch` package

In [`autoware.launch.xml`](https://github.com/autowarefoundation/autoware_launch/blob/main/autoware_launch/launch/autoware.launch.xml) of the `autoware_launch` package, parameters in the `autoware_launch` package (`Launch Parameter`) are used.
For example, the `tier4_planning_launch` package receives `$(find-pkg-share autoware_launch)/config/tier4_planning_launch` as the `tier4_planning_launch_param_path` argument as follows.

```xml
<include file="$(find-pkg-share tier4_planning_launch)/launch/planning.launch.xml">
  <arg name="tier4_planning_launch_param_path" value="$(find-pkg-share autoware_launch)/config/tier4_planning_launch"/>
</include>
```

> ### When launching autoware via the component launch

Reviewer 2: ```suggestion
### When launching an individual component via the `tier4_COMPONENT_launch`
```
How about this?

Author: https://github.com/autowarefoundation/autoware-documentation/pull/171/commits/9ebdfe40092fcbd08b2741dccbc7dfce0a57afff
DONE. I remained the name component launch.

In [`planning.launch.xml`](https://github.com/autowarefoundation/autoware.universe/blob/main/launch/tier4_planning_launch/launch/planning.launch.xml) of the `tier4_planning_launch` package, the default variable for the `tier4_planning_launch_param_path` argument is `$(find-pkg-share tier4_planning_launch)/config`.
Therefore, parameters in the `tier4_planning_launch` package (`Component Parameter`) are used.

```xml
<arg name="tier4_planning_launch_param_path" default="$(find-pkg-share tier4_planning_launch)/config" description="tier4_planning_launch parameter path"/>
```

![parameters-to-be-used](images/parameters-to-be-used.svg){: style="width:1200px"}

## Automatic synchronization of parameters

> There is a PR generated automatically to synchronize `Launch Parameter` with `Component Parameter`.

Author: https://github.com/autowarefoundation/autoware-documentation/pull/171/commits/9ebdfe40092fcbd08b2741dccbc7dfce0a57afff
DONE

You can see the synchronization setting file [here](https://github.com/autowarefoundation/autoware_launch/blob/main/.github/sync-param-files.yaml).
Note that all parameter files are listed one by one.

![parameter-sync](images/parameter-sync.svg){: style="width:900px"}

## How to manage parameters

We explain how to manage parameters in some use cases.

### When creating a new package

- Locates the parameters of `Node Parameter` and `Component Parameter` for the new package in `autowarefoundation/autoware.universe`.
- Create a PR to `autowarefoundation/autoware.universe`.
- Update [a synchronization setting file](https://github.com/autowarefoundation/autoware_launch/blob/main/.github/sync-param-files.yaml).
- Sync PR to `autowarefoundation/autoware_launch` is created automatically.
  - Or you can create this PR manually.

> - Merge both PRs.

Author: https://github.com/autowarefoundation/autoware-documentation/pull/171/commits/9ebdfe40092fcbd08b2741dccbc7dfce0a57afff
DONE

> - Merge both PRs.

Reviewer 2: Thanks! 
Seems I should have added an empty line before "Step 2.". Would you fix this too?
![image](https://user-images.githubusercontent.com/44218668/192417060-ed53b293-2d65-4f24-9431-ad89e5682607.png)

> - Merge both PRs.

Author: Sorry
https://github.com/autowarefoundation/autoware-documentation/pull/171/commits/3df5f25bacffc57448379cf61ad555fe2ac86317

### When modifying general parameters

- Modify the parameters of `Node Parameter` and `Component Parameter` in `autowarefoundation/autoware.universe`.
- Create a PR to `autowarefoundation/autoware.universe`.
- If you add a new config file, update [a synchronization setting file](https://github.com/autowarefoundation/autoware_launch/blob/main/.github/sync-param-files.yaml).
- Sync PR to `autowarefoundation/autoware_launch` is created automatically.
  - Or you can create this PR manually.

> - Merge both PRs.

Author: https://github.com/autowarefoundation/autoware-documentation/pull/171/commits/9ebdfe40092fcbd08b2741dccbc7dfce0a57afff
DONE

> - Merge both PRs.

Reviewer 2: Same as https://github.com/autowarefoundation/autoware-documentation/pull/171#discussion_r980670395

> - Merge both PRs.

Author: https://github.com/autowarefoundation/autoware-documentation/pull/171/commits/3df5f25bacffc57448379cf61ad555fe2ac86317

> ### When modifying integrated parameters
> 
> Note: Assuming that there are your forked repositories of `autowarefoundation/autoware` and `autowarefoundation/autoware_launch`.
> 
> This change does not affect the Autoware Foundation's repository.
> Therefore, just modifying your forked repository is fine.
> 
> - Modify the parameters of `Launch Parameter` in your forked `autoware_launch` repository.
> - Create a PR and merge.

Reviewer 2: Is this what you've intended?
```suggestion
### How to maintain your custom parameters

For example if you want to integrate Autoware with your own vehicles, you may need to tune some parameters in Autoware.

A recommended way is to use `Launch Parameter` in `autowarefoundation/autoware_launch` as your custom parameters. Please fork the repository (e.g. as `YOUR_ACCOUNT/autoware_launch.YOURS`) and modify the parameters as you want.

One advantage of doing so is that the `sync-param-files` GitHub Actions helps you to keep `YOUR_ACCOUNT/autoware_launch.YOURS` up-to-date with `autowarefoundation/autoware.universe`. The GitHub Actions workflow will automatically create sync PR on a regular basis, so you can incorporate the latest changes in `autowarefoundation/autoware.universe` when you want.
```

Author: https://github.com/autowarefoundation/autoware-documentation/pull/171/commits/9ebdfe40092fcbd08b2741dccbc7dfce0a57afff
DONE

I really appreciate your better expressions!
